### PR TITLE
Feature/user profile

### LINKED ISSUE #1434: Tutor profile - Display basic tutor profile details
### Link task to story

Story: https://github.com/Open-Study-College/osc/issues/1419

### Description

n/a

### LINKED ISSUE #1539: Add and style user profile picture
### Link task to story

Story: https://github.com/Open-Study-College/osc/issues/1507

### Description

n/a

### LINKED ISSUE #1540: Add user profile form - Tutor
### Link task to story

Story: https://github.com/Open-Study-College/osc/issues/1507

### Description

n/a

### LINKED ISSUE #1541: Add additional info form - tutor
### Link task to story

Story: https://github.com/Open-Study-College/osc/issues/1507

### Description

n/a

### LINKED ISSUE #1547: Student view - Layout student profile
### Link task to story

Story: https://github.com/Open-Study-College/osc/issues/1507

### Description

n/a

### LINKED ISSUE #1548: Student profile - Add user avatar
### Link task to story

Story: https://github.com/Open-Study-College/osc/issues/1507

### Description

n/a

### LINKED ISSUE #1549: Student profile - Update user details form
### Link task to story

Story: https://github.com/Open-Study-College/osc/issues/1507

### Description

n/a

### LINKED ISSUE #1550: Student profile Update additional info farm
### Link task to story

Story: https://github.com/Open-Study-College/osc/issues/1507

### Description

n/a

### PR DESCRIPTION
**Pull request by stevenbister**
### 🧹 Please check off the following before submitting:

- [x] Close issues
- [x] Request reviewers
- [x] Assign yourself
- [ ] Add labels (e.g. `build:osc-ecommerce`)
- [x] Add project
- [x] Set project status
- [x] Add milestone

---

## ☑ List all github issues that this PR will close:

- Closes #1434
- Closes #1539
- Closes #1540 
- Closes #1541 
- Closes #1547 
- Closes #1548 
- Closes #1549 
- Closes #1550 

## 📝 Description

Adds the profile routes and profile form components for admins and tutors to be able to edit and update their data. 
This is just a visual pr so it's mostly mock data and there is no functionality set up as of yet.

### Tutor
localhost_3000_profile_details
localhost_3000_profile_details (1)

### Admin
localhost_3000_profile_details (2)
localhost_3000_profile_details (3)

### Student
localhost_3000_profile_details (4)
localhost_3000_profile_details (5)

I've not added the courses for students yet as I believe they are still being worked on and could be big enough to warrant their own pr anyway.

I've got a PR open #1618 which creates a proper component for the navbar in here, which I'll replace once it's merged. So don't worry about that file too much as it's mostly a copy from there.

## ⛳️ Current behaviour (updates)

N/a

## 🚀 New behaviour

- Adds index route to profile so we can redirect to detials
- Adds details and additional info forms
- Adds user avatar
- Export datatable from osc-ui and adds to tutor details
- Mocks data
- Minor updates to Avatar and Badge components so we can handle the needs for the profile picture.

## 💣 Is this a breaking change (Yes/No): No

> Please describe any breaking changes.

## 🧾 Final Checklist

- [ ] I've considered possible cross-platform impacts (Mac, Linux, Windows)
- [x] I've considered possible cross-device impacts (iOS, Android)
- [x] I've considered possible cross-browser impacts (Chrome, Safari, Edge, Firefox)
- [ ] I've added tests to cover my changes (Unit, E2E, Integration, Contract, Functional, Acceptance etc)
- [ ] I've added or updated any documentation.

## 📝 Additional Information

Something to note. As we've got a few PRs in review this probably won't work fully / will have errors until #1596 & #1610 are merged and this is rebased against main. In the mean time you might have to grab the missing functions and drop them in where they're needed

### Summary by CodeRabbit

- New Feature: Introduced `NavLink` and `NavLinkList` components for improved navigation with active state handling.
- New Feature: Added `AdditionalInfoForm`, `UserDetailsForm`, and `TutorCoursesTable` components to capture more user information and display course details.
- New Feature: Implemented `ProfileAvatar` component for displaying and editing user avatars.
- New Feature: Created `NavBar` component for dynamic navigation based on user roles.
- Refactor: Updated `root.tsx` to include `I18nProvider` for consistent locale specification between server and client.
- New Feature: Added `objectsAreDifferent` utility function for comparing two objects.
- Style: Updated styles in `osc-ui` package to provide more flexibility in badge positioning and size.
- Test: Added new test suite for the `objectsAreDifferent` function.

### CODE REVIEW

**Review by jkellerman — APPROVED** [review]
LGTM

File: packages/osc-academic-hub/app/root.tsx [hunk 1]
```diff
@@ -150,11 +151,14 @@ export default function App() {
 
     return (
         <Document>
-            <SpritesheetProvider spriteSheetPath={spritesheet}>
-                <SkipLink anchor="main-content">Skip to main content</SkipLink>
-
-                <Outlet />
-            </SpritesheetProvider>
+            {/* Recommended to always specify locale to ensure server and client match - https://react-spectrum.adobe.com/react-aria/internationalization.html*/}
+            <I18nProvider locale="en-GB">
```
Comment (mattwendzina) [inline packages/osc-academic-hub/app/root.tsx:155]: Wondering whether we're going to need to do a check here for the locale of a user as they may not be in the UK? 🤔

**Review by mattwendzina — COMMENTED** [review]
This is looking ace mate 🙌  
Got a few comments in terms of functionality. I guess some of them might just get addressed as more work gets done on it, so apologies in advance for that!

- In Desktop view the Postcode is getting cut off and is having to scroll. Would be better if it was bigger and you could see the entire postcode:
Screenshot 2023-09-27 at 10 28 02

- Might be good for save changes button to only be visible/disabled unless they've actually made a change. Also when you hit 'Save Changes' the 'Reset' button appears which seems a bit strange - in part because you can hit 'Save Changes' when you haven't made any changes, so there's nothing to reset. 
- When you do make a change the reset button appears. If you click it it removes the changes, but maybe then worth hiding the reset button again?

- There's quite a lot of errors in the console atm which is making it hard to know if any are related to a specific PR or not, but wondering if this one might be?

```
Warning: Function components cannot be given refs. Attempts to access this ref will fail. Did you mean to use React.forwardRef()?
Check the render method of `SlotClone`.
```

- The size of the profile card increases when you go to courses because there isn't a scrollbar (only on desktop). Feel like there is an easy css fix that stops the size from changing when the scrollbar isn't visible, although I don't remember what that is, so I may be making that up lol. If there is though/it's quick it might be nice to add. 

- Should the drawer be sticky so that it can be viewed wherever a user is on the page? I was scrolled down when I first clicked in and couldn't see anything, see screenshot. 

Screenshot 2023-09-27 at 10 35 43

**Comment by stevenbister** [comment]
> This is looking ace mate 🙌 Got a few comments in terms of functionality. I guess some of them might just get addressed as more work gets done on it, so apologies in advance for that!
> 
> * In Desktop view the Postcode is getting cut off and is having to scroll. Would be better if it was bigger and you could see the entire postcode:
> 
> Screenshot 2023-09-27 at 10 28 02

Ooo yeah good spot, sorted!

> * Might be good for save changes button to only be visible/disabled unless they've actually made a change. Also when you hit 'Save Changes' the 'Reset' button appears which seems a bit strange - in part because you can hit 'Save Changes' when you haven't made any changes, so there's nothing to reset.
> * When you do make a change the reset button appears. If you click it it removes the changes, but maybe then worth hiding the reset button again?

I disagree with disabling the save button, I can't find the article now but I read somewhere that it's not great for a11y to disable submit buttons on forms really, as they lose context to screen readers. And having it active I don't think really matters personally.

I have simplified my compare objects function so that should be more consistent in returning whether the objects match, which re-hides the button after resetting.

> * There's quite a lot of errors in the console atm which is making it hard to know if any are related to a specific PR or not, but wondering if this one might be?
> 
> ```
> Warning: Function components cannot be given refs. Attempts to access this ref will fail. Did you mean to use React.forwardRef()?
> Check the render method of `SlotClone`.
> ```

Yeah these all occur on ecommerce as well, I think there's an issue with one of the items in the modal/drawer that uses Slot. Should be a quick fix, just need to get around to it lol

> * The size of the profile card increases when you go to courses because there isn't a scrollbar (only on desktop). Feel like there is an easy css fix that stops the size from changing when the scrollbar isn't visible, although I don't remember what that is, so I may be making that up lol. If there is though/it's quick it might be nice to add.

Yes there is! `scrollbar-gutter: stable` -- added

> * Should the drawer be sticky so that it can be viewed wherever a user is on the page? I was scrolled down when I first clicked in and couldn't see anything, see screenshot.
> 
> Screenshot 2023-09-27 at 10 35 43

Aha yeah that should be stuck to the top, I've opened an issue to fix this next sprint https://github.com/Open-Study-College/osc/issues/1659

**Comment by mattwendzina** [comment]
Nice mate - changes all look good! Fair enough on the save changes didn't know about the a11y concerns with disabled buttons 👍 

Only one other thing I'd say which may be annoying and doesn't necessarily have to be dealt with but I think is a helpful part of the UX. If you make changes and then change tab and then come back you lose all your changes. I appreciate it when I get a warning that leaving this page/tab without saving changes mean they will be lost, or that the data isn't lost. I guess given the amount of data - ie just a bit of profile data it isn't that big of an issue 🤷 

This probably isn't a 5 minute job and is just my opinion on this so happy to approve as is if you disagree/open a new ticket if you agree but can be done further down the line.

**Review by mattwendzina — COMMENTED** [review]
Changes look good mate 👌 

Just let one final comment, happy to approve either way though.

**Comment by stevenbister** [comment]
> Nice mate - changes all look good! Fair enough on the save changes didn't know about the a11y concerns with disabled buttons 👍
> 
> Only one other thing I'd say which may be annoying and doesn't necessarily have to be dealt with but I think is a helpful part of the UX. If you make changes and then change tab and then come back you lose all your changes. I appreciate it when I get a warning that leaving this page/tab without saving changes mean they will be lost, or that the data isn't lost. I guess given the amount of data - ie just a bit of profile data it isn't that big of an issue 🤷
> 
> This probably isn't a 5 minute job and is just my opinion on this so happy to approve as is if you disagree/open a new ticket if you agree but can be done further down the line.

Yeah I know what you mean, I would like to do it but unfortunately there's no way to intercept routes (afaik) in Remix at the moment. So I could show a modal if there are changes when a user clicks on a nav link but it would still navigate the page so you'd end up on `/additional-info` but have a modal talking about the form on `/details` (if that makes sense) :/ Unless you've got any ideas?

**Comment by mattwendzina** [comment]
> > Nice mate - changes all look good! Fair enough on the save changes didn't know about the a11y concerns with disabled buttons 👍
> > Only one other thing I'd say which may be annoying and doesn't necessarily have to be dealt with but I think is a helpful part of the UX. If you make changes and then change tab and then come back you lose all your changes. I appreciate it when I get a warning that leaving this page/tab without saving changes mean they will be lost, or that the data isn't lost. I guess given the amount of data - ie just a bit of profile data it isn't that big of an issue 🤷
> > This probably isn't a 5 minute job and is just my opinion on this so happy to approve as is if you disagree/open a new ticket if you agree but can be done further down the line.
> 
> Yeah I know what you mean, I would like to do it but unfortunately there's no way to intercept routes (afaik) in Remix at the moment. So I could show a modal if there are changes when a user clicks on a nav link but it would still navigate the page so you'd end up on `/additional-info` but have a modal talking about the form on `/details` (if that makes sense) :/ Unless you've got any ideas?

Lol, yeah that wouldn't really be ideal. Nothing off the top of my head tbh. Maybe we leave it for now. I think it would be more of an issue if it were pages where there was potentially a lot of data entry but given how minimal it is here I don't think it's too big of an issue. 

Happy to approve 👍

**Review by mattwendzina — APPROVED** [review]
LGTM

**Pull request merged**